File: Retos/Reto #35 - PRIMEROS PASOS [Fácil]/clojure/ingSinTitulo.clj
;; Punto 1: Hola, mundo!
(println "Hola, mundo!")

;; Punto 2: Crea una variable de texto o string
(def miTexto "¡Hola desde Clojure!")

;; Punto 3: Crea una variable de número entero
(def miEntero 42)

;; Punto 4: Crea una variable de número con decimales
(def miDecimal 3.14)

;; Punto 5: Crea una variable de tipo booleano
(def miBooleano true)

;; Punto 6: Crea una constante (en Clojure, las constantes se definen con def)
(def MI_CONSTANTE 10)

;; Punto 7: Usa un if, else if y else
(if (> miEntero 50)
  (println "El número es mayor que 50")
  (if (< miEntero 50)
    (println "El número es menor que 50")
    (println "El número es igual a 50")))

;; Punto 8: Crea un vector
(def miVector [1 2 3 4 5])

;; Punto 9: Crea una lista
(def miLista '("Manzana" "Banana" "Naranja"))

;; Punto 10: Crea una tupla (no aplicable en Clojure)

;; Punto 11: Crea un set
(def miSet #{ "Rojo" "Verde" "Azul" })

;; Punto 12: Crea un diccionario (map en Clojure)
(def miMapa { :clave1 "valor1" :clave2 "valor2" })

;; Punto 13: Usa un ciclo for (no es común en Clojure, se prefiere el uso de funciones de alto nivel)
(doseq [x miVector]
  (println x))

;; Punto 14: Usa un ciclo foreach (no es común en Clojure, se prefiere el uso de funciones de alto nivel)
(doseq [elemento miLista]
  (println elemento))

;; Punto 15: Usa un ciclo while (no es común en Clojure, se prefiere el uso de funciones de alto nivel)
(defn loop-ciclo [contador]
  (if (< contador 3)
    (do
      (println (str "Contador: " contador))
      (recur (inc contador)))))

(loop-ciclo 0)

;; Punto 16: Crea una función sin parámetros que no retorne nada
(defn funcion-sin-parametros []
  (println "Función sin parámetros"))

;; Punto 17: Crea una función con parámetros que no retorne nada
(defn funcion-con-parametros [param1 param2]
  (println (str "Parámetro 1: " param1))
  (println (str "Parámetro 2: " param2)))

;; Punto 18: Crea una función con parámetros que retorne valor
(defn funcion-con-retorno [a b]
  (+ a b))

;; Punto 19: Crea una clase (en Clojure se trabaja mayormente con estructuras de datos inmutables)
(defrecord Persona [nombre edad])

;; Punto 20: Muestra control de excepciones (se manejan con try y catch)
(try
  (/ miEntero 0)
  (catch Exception e
    (println (str "Error: " (.getMessage e)))))

File: Retos/Reto #35 - PRIMEROS PASOS [Fácil]/perl/ingSinTitulo.pl
#!/usr/bin/perl

use strict;
use warnings;

# Punto 1: Hola, mundo!
print "Hola, mundo!\n";

# Punto 2: Crea una variable de texto o string
my $miTexto = "¡Hola desde Perl!\n";

# Punto 3: Crea una variable de número entero
my $miEntero = 42;

# Punto 4: Crea una variable de número con decimales
my $miDecimal = 3.14;

# Punto 5: Crea una variable de tipo booleano
my $miBooleano = 1;  # Verdadero

# Punto 6: Crea una constante (no aplicable en Perl)

# Punto 7: Usa un if, else if y else
if ($miEntero > 50) {
    print "El número es mayor que 50\n";
} elsif ($miEntero < 50) {
    print "El número es menor que 50\n";
} else {
    print "El número es igual a 50\n";
}

# Punto 8: Crea un Array (no aplicable en Perl)

# Punto 9: Crea una lista (no aplicable en Perl)

# Punto 10: Crea una tupla (no aplicable en Perl)

# Punto 11: Crea un set (no aplicable en Perl)

# Punto 12: Crea un diccionario (no aplicable en Perl)

# Punto 13: Usa un ciclo for
foreach my $elemento (1..5) {
    print "$elemento\n";
}

# Punto 14: Usa un ciclo foreach
foreach my $elemento ("Manzana", "Banana", "Naranja") {
    print "$elemento\n";
}

# Punto 15: Usa un ciclo while
my $contador = 0;
while ($contador < 3) {
    print "Contador: $contador\n";
    $contador++;
}

# Punto 16: Crea una función sin parámetros que no retorne nada
sub funcionSinParametros {
    print "Función sin parámetros\n";
}
funcionSinParametros();

# Punto 17: Crea una función con parámetros que no retorne nada
sub funcionConParametros {
    my ($param1, $param2) = @_;
    print "Parámetro 1: $param1\n";
    print "Parámetro 2: $param2\n";
}
funcionConParametros(1, "dos");

# Punto 18: Crea una función con parámetros que retorne valor
sub funcionConRetorno {
    my ($a, $b) = @_;
    return $a + $b;
}
my $resultado = funcionConRetorno(3, 4);
print "Resultado: $resultado\n";

# Punto 19: Crea una clase (no hay clases en Perl, se usan paquetes)

# Punto 20: Muestra control de excepciones (eval en Perl)
eval {
    my $division = $miEntero / 0;
    print "$division\n";
};
if ($@) {
    print "Error: $@\n";
}
